File: api/src/routes/history.ts
import { Router, Request, Response } from 'express';
import { readFile } from 'fs/promises';
import { join } from 'path';
import { homedir } from 'os';

const router = Router();

function getClaudeDir(): string {
  return join(homedir(), '.claude');
}

interface HistoryEntry {
  display: string;
  timestamp: number;
  project?: string;
  pastedContents?: Record<string, unknown>;
}

// GET /history - Get prompt history
router.get('/', async (req: Request, res: Response) => {
  try {
    const claudeDir = getClaudeDir();
    const historyPath = join(claudeDir, 'history.jsonl');

    const limit = Math.min(parseInt(req.query.limit as string) || 50, 100);
    const offset = parseInt(req.query.offset as string) || 0;
    const projectFilter = req.query.project as string | undefined;
    const searchQuery = req.query.search as string | undefined;
    const startDate = req.query.startDate ? new Date(req.query.startDate as string).getTime() : undefined;
    const endDate = req.query.endDate ? new Date(req.query.endDate as string).getTime() : undefined;

    const content = await readFile(historyPath, 'utf-8');
    const lines = content.trim().split('\n').filter(l => l.trim());

    let entries: HistoryEntry[] = [];

    for (const line of lines) {
      try {
        const parsed = JSON.parse(line) as HistoryEntry;
        entries.push(parsed);
      } catch {
        // Skip malformed lines
      }
    }

    // Apply filters
    if (projectFilter) {
      entries = entries.filter(e => e.project?.includes(projectFilter));
    }

    if (searchQuery) {
      const query = searchQuery.toLowerCase();
      entries = entries.filter(e => e.display?.toLowerCase().includes(query));
    }

    if (startDate) {
      entries = entries.filter(e => e.timestamp >= startDate);
    }

    if (endDate) {
      entries = entries.filter(e => e.timestamp <= endDate);
    }

    // Sort by timestamp descending (most recent first)
    entries.sort((a, b) => b.timestamp - a.timestamp);

    // Paginate
    const total = entries.length;
    const paginatedEntries = entries.slice(offset, offset + limit);

    res.json({
      data: paginatedEntries,
      meta: {
        total,
        limit,
        offset,
        hasMore: offset + limit < total
      }
    });
  } catch {
    res.json({
      data: [],
      meta: {
        total: 0,
        limit: 50,
        offset: 0,
        hasMore: false
      }
    });
  }
});

export { router as historyRouter };

File: api/src/routes/history.py
"""History routes for Claude Explorer API.

Provides access to prompt history - a chronological log of all user
prompts across projects without full conversation context.
"""

from fastapi import APIRouter, Query

from ..models import HistoryResponse
from ..utils import encode_project_path, get_claude_dir, parse_jsonl_file

router = APIRouter(prefix="/history", tags=["history"])


@router.get("/", response_model=HistoryResponse)
async def get_history(
    project: str | None = Query(
        None,
        description="Filter to prompts from projects containing this path substring"
    ),
    start_date: str | None = Query(
        None,
        alias="startDate",
        description="Filter prompts on or after this date (ISO 8601 format, e.g., '2024-01-15')"
    ),
    end_date: str | None = Query(
        None,
        alias="endDate",
        description="Filter prompts on or before this date (ISO 8601 format)"
    ),
    search: str | None = Query(
        None,
        description="Filter prompts containing this text (case-insensitive)"
    ),
    limit: int = Query(
        50,
        le=100,
        description="Maximum number of prompts to return (max 100)"
    ),
    offset: int = Query(
        0,
        ge=0,
        description="Number of prompts to skip for pagination"
    ),
) -> HistoryResponse:
    """Get prompt history across all sessions.

    Returns history entries sorted by timestamp descending (most recent
    first). Each entry contains the prompt text, timestamp, project path,
    and any pasted content.

    This provides a quick activity timeline without loading full
    session transcripts.

    Returns:
        data: List of HistoryEntry objects
        meta: Pagination metadata with total, limit, offset, hasMore
    """
    claude_dir = get_claude_dir()
    history_path = claude_dir / "history.jsonl"

    if not history_path.exists():
        return {
            "data": [],
            "meta": {"total": 0, "limit": limit, "offset": offset, "hasMore": False},
        }

    try:
        content = history_path.read_text()
        entries = parse_jsonl_file(content)

        # Apply filters
        if project:
            entries = [e for e in entries if project in e.get("project", "")]

        if search:
            query = search.lower()
            entries = [e for e in entries if query in e.get("display", "").lower()]

        if start_date:
            from datetime import datetime
            start_ts = datetime.fromisoformat(start_date).timestamp() * 1000
            entries = [e for e in entries if e.get("timestamp", 0) >= start_ts]

        if end_date:
            from datetime import datetime
            end_ts = datetime.fromisoformat(end_date).timestamp() * 1000
            entries = [e for e in entries if e.get("timestamp", 0) <= end_ts]

        # Sort by timestamp descending
        entries.sort(key=lambda e: e.get("timestamp", 0), reverse=True)

        # Paginate
        total = len(entries)
        paginated = entries[offset : offset + limit]

        # Transform entries to use projectPath and projectId
        transformed = []
        for entry in paginated:
            project = entry.get("project")
            transformed.append({
                "display": entry.get("display", ""),
                "timestamp": entry.get("timestamp", 0),
                "projectPath": project,
                "projectId": encode_project_path(project) if project else None,
                "pastedContents": entry.get("pastedContents"),
            })

        return {
            "data": transformed,
            "meta": {
                "total": total,
                "limit": limit,
                "offset": offset,
                "hasMore": offset + limit < total,
            },
        }
    except Exception:
        return {
            "data": [],
            "meta": {"total": 0, "limit": limit, "offset": offset, "hasMore": False},
        }
